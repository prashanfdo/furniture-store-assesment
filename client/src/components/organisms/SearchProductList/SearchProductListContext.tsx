import constate from "constate";
import { useQuery, gql } from "@apollo/client";
import { Article } from "types";
type GetSearchProductListResponse = {
  searchData: {
    total: number;
    articles: Article[];
  };
};
const GET_PRODUCT_LIST = gql`
  query GetSearchProductList($query: String!, $locale: Locale!) {
    searchData: search(locale: $locale, query: $query) {
      ... on SearchResultArticles {
        total
        articles {
          name
          variantName
          images(format: WEBP, maxWidth: 200, maxHeight: 200, limit: 4) {
            path
            type
          }
          prices {
            currency
            regular {
              value
            }
            special {
              value
              discount
            }
          }
        }
      }
    }
  }
`;

type SearchProductListContextProviderProps = {
  searchText: string;
  page: number;
};
const [SearchProductListContextProvider, useSearchProductListContext] = constate(
  ({ searchText, page }: SearchProductListContextProviderProps) => {
    const { loading, error, data } = useQuery<GetSearchProductListResponse>(GET_PRODUCT_LIST, {
      variables: { query: searchText, locale: "de_DE" },
    });
    console.log(data?.searchData);
    return {
      searchText,
      page,
      loading,
      error,
      data: data?.searchData,
    };
  }
);

export { SearchProductListContextProvider, useSearchProductListContext };
