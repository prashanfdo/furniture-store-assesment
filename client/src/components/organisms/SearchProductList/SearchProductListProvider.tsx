import constate from "constate";
import useGetSearchProductList from "queries/useGetSearchProductList";

type SearchProductListContextProviderProps = {
  searchText: string;
  page: number;
};
const [SearchProductListProvider, useSearchProductListContext] = constate(
  ({ searchText, page }: SearchProductListContextProviderProps) => {
    const { loading, error, data } = useGetSearchProductList(searchText);
    return {
      searchText,
      page,
      loading,
      error,
      data: data?.searchData,
    };
  },
);

export { useSearchProductListContext };

export default SearchProductListProvider;
