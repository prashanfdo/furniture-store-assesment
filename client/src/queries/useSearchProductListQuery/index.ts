import { useQuery } from "@apollo/client";
import SEARCH_PRODUCT_LIST_QUERY from "./query";
import { SearchProductListQueryData, SearchProductListQueryVars } from "./useSearchProductListQuery.types";
import { DEFAULT_LOCALE } from "helpers/constants";

export default function useSearchProductListQuery(searchText: string) {
  return useQuery<SearchProductListQueryData, SearchProductListQueryVars>(SEARCH_PRODUCT_LIST_QUERY, {
    variables: { query: searchText, locale: DEFAULT_LOCALE },
  });
}
