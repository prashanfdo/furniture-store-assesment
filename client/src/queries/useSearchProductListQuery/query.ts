import { gql } from "@apollo/client";

const SEARCH_PRODUCT_LIST_QUERY = gql`
  query GetSearchProductList($query: String!, $locale: Locale!) {
    searchData: search(locale: $locale, query: $query) {
      ... on SearchResultArticles {
        total
        articles {
          name
          variantName
          images(format: WEBP, maxWidth: 400, maxHeight: 400, limit: 4) {
            path
            type
          }
          prices {
            currency
            regular {
              value
            }
            special {
              value
              discount
            }
          }
        }
      }
    }
  }
`;

export default SEARCH_PRODUCT_LIST_QUERY;
