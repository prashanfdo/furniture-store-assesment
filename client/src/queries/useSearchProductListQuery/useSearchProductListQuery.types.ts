import { Article } from "types";

export type SearchProductListQueryData = {
  searchData: {
    total: number;
    articles: Article[];
  };
};

export type SearchProductListQueryVars = {
  query: string;
  locale: string;
};
