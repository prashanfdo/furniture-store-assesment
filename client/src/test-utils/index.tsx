import { render, RenderOptions, renderHook, RenderHookOptions, RenderHookResult } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import user from "@testing-library/user-event";
import { ProvidersForTests } from "providers";

type GraphQlMocks = React.ComponentProps<typeof ProvidersForTests>["mocks"];

type WrapperProps = React.ComponentProps<typeof ProvidersForTests> & {
  children: React.ReactNode;
};
const Wrapper: FC<WrapperProps> = ({ children, mocks, initialRoute }: WrapperProps) => {
  return (
    <ProvidersForTests mocks={mocks} initialRoute={initialRoute}>
      {children}
    </ProvidersForTests>
  );
};
const wrapper =
  (mocks: GraphQlMocks, initialRoute?: string) =>
  ({ children }: { children: React.ReactNode }) =>
    (
      <Wrapper mocks={mocks} initialRoute={initialRoute}>
        {children}
      </Wrapper>
    );

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & { gqlMocks: GraphQlMocks; initialRoute: string },
) => {
  const view = render(ui, { wrapper: wrapper(options?.gqlMocks, options?.initialRoute), ...options });
  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};

export const customRenderHook = <T, R>(
  callback: (props: T) => R,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { gqlMocks?: GraphQlMocks; initialRoute?: string },
): RenderHookResult<R, T> => {
  return renderHook(callback, { ...options, wrapper: wrapper(options?.gqlMocks, options?.initialRoute) });
};

export { faker } from "@faker-js/faker";
export * from "@testing-library/react";
export * from "./generate-data";
export { customRender as render };
export { customRenderHook as renderHook };
export { user };
