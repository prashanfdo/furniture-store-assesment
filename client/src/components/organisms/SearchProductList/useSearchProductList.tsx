import useSearchProductListQuery from "queries/useSearchProductListQuery";

const useSearchProductList = (searchText: string, page: number) => {
  const { loading, error, data } = useSearchProductListQuery(searchText);
  return {
    loading,
    error,
    data: data?.searchData,
  };
};

export default useSearchProductList;
