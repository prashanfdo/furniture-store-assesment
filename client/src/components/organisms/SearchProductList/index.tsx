import { ArticleListHeader } from "components/atoms";
import { ArticleGrid } from "components/molecules";
import useSearchProductList from "./useSearchProductList";

type SearchProductListProps = {
  searchText: string;
  page: number;
};
const SearchProductList: React.FC<SearchProductListProps> = ({ searchText, page }: SearchProductListProps) => {
  const { data } = useSearchProductList(searchText, page);

  if (data === undefined) {
    return null;
  }
  return (
    <div className="grid grid-rows-[auto_1fr] py-2 px-8">
      <ArticleListHeader name={`Search Results for "${searchText}"`} articleCount={data.total} />
      <ArticleGrid products={data.articles as any} />
    </div>
  );
};

export default SearchProductList;
