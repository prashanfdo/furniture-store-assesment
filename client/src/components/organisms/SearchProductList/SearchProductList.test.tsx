// TODO: clean all below like imports, must import from 'test-utils
import useSearchProductListQuery from "queries/useSearchProductListQuery";
import {
  SearchProductListQueryData,
  SearchProductListQueryVars,
} from "queries/useSearchProductListQuery/useSearchProductListQuery.types";
import { createArticleList, createGqlResult, render } from "test-utils";
import SearchProductList from "./index";

jest.mock("components/atoms", () => ({
  ArticleListHeader: () => <div data-testid="article-list-header" />,
}));
jest.mock("components/molecules", () => ({
  ArticleGrid: () => <div data-testid="article-grid" />,
}));
jest.mock("queries/useSearchProductListQuery");
const mockUseSearchProductListQuery = useSearchProductListQuery as jest.MockedFunction<
  typeof useSearchProductListQuery
>;

describe("SearchProductList", () => {
  it("should render", () => {
    mockUseSearchProductListQuery.mockReturnValueOnce(
      createGqlResult<SearchProductListQueryData, SearchProductListQueryVars>({
        data: {
          searchData: {
            articles: createArticleList(10),
            total: 10,
          },
        },
      }),
    );
    const { container } = render(<SearchProductList searchText="test" page={1} />);
    expect(container).toMatchSnapshot();
    expect(mockUseSearchProductListQuery).toHaveBeenCalledTimes(1);
  });

  // TODO: add tests for loading and error states
});
