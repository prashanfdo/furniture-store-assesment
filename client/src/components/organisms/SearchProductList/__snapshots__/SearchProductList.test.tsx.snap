// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchProductList should render 1`] = `
<div>
  <div
    class="grid grid-rows-[auto_1fr] py-2 px-8"
  >
    <div
      data-testid="article-list-header"
    />
    <div
      data-testid="article-grid"
    />
  </div>
</div>
`;
